Nodes records have now authentication. Adapted to Ethereum Node record scheme.

MaximilianoRey94 opened the pull request:
# Description
One of the Kademelia implementation TODO indicates Kademlia Node Records, the message object that binds a Kademlia ID with his direction in the sub-net, should be signed, using the model of Ethereum Node Records. This model specify the secp256k1 encryption system must be used.

This system is important because without any identification, everyone can send a fake node record binding an ID of another node with an IP that he controls. Later, it is possible to mount a malicious node on the web with this ID, who sends fake information causing the web doesn’t trust in the ID and the original node will be beaned.

Also  a user can create malicious nodes with IDs similar to a target node and monopolizes its connection with the net.

# Proposed Solution
In this implementation a node’s Kademlia ID is its encoded secp256k1 public key. With this, we solved the problem of binding a public key with a direction on the web, because if a node changes its public key, it changes its direction.

The node record has now a sign of the rest of the node record information, using SHA-256 to make the hash. When a node receives a node record, verify if the sign is valid, and if it’s not, the node record will not add.

Also, it's now impossible to generate nodes IDs arbitrarily, so is more difficult to monopolizes a node connection how is described before.

Additionally, a sequence number has been added. This number increment himself everytime the node refresh, and should increment manually if the node shutdown and later is online again with other IP. If there are two record nodes with the same vector ID, the system only register the node record with the higher sequence number. This prevent a node to register obsolete information.

# Important Changes Introduced
- NodeRecord subclass have now a sign. KRouter's node's id is a 33 byte public key
- Add package crypto
- Add spongycastle to dependencies
- Add StreamObject file
- A crypto package was added. It contains the functions to generate key pairs, sign and verify a signature
- There is now a object NodeRecord with functions for create signed NodeRecords and verify the signature of a NodeRecord. It required a Codec[A]. 
- Krouter now must receive a Codec[A] and the private key, because internally required verify the discovered NodeRecords and create new NodeRecords with his ID and other sequence number of the original
- Generators have been alterated for generate NodeRecords with a valid signature. Also defaultBitLength have been changed to 264
- Nodes now don't accept NodeRecords with a invalid signature or a sequence number lower than the last register they have about this ID (if it have any register)

## Problems
- Kademlia specification says the Node’s ID has 160 bits. In the implementation, this can be different, because nodes only check that a node ID has the same length of its own ID. So, all nodes of the web must have IDs with the same length, that can be or not of 160 bits. With this change, the length of nodes ID must be of 264 bits.
- You can’t change your public key without losing your identity. In the blockchain case this is not very important, because there are no funds associated with this. A node ID is only an identification, and a node can change it without important repercussions. If a node is hacked and its private key is stolen, the node must change his ID.

# Testing
There are testing added in KRouterSpec that verify:
- A node accepts a NodeRecord with a valid signature
- A node doesn't accept a NodeRecord with a invalid signature.
- If a node receive two NodeRecords with valid signature and different sequence number, he store the record with the higher sequence number

KonradStaniec committed:
[PM-853] Various improvements to krouter

KonradStaniec committed:
[PM-853] Pr comments. Add router state handling via ref

KonradStaniec committed:
[PM-853] Fix typos. Add auxiliary function.

KonradStaniec committed:
[PM-854] Pr comments

Replace observable by scheduler
Fix flaky tests in krouterspec

KonradStaniec committed:
[PM-854] Bump version

KonradStaniec committed:
[PM-1028] Simple kademlia refresh process

KonradStaniec committed:
[PM-1028] Pr comments. Add more documentation to config and kbuckets

KonradStaniec committed:
[PM-1078] Add integration test for kademlia discovery

KonradStaniec committed:
[PM-1078] Shorten the integration test names

KonradStaniec committed:
[PM-1078] Fix nullpointer in ConnectableSubject

Commit:
Nodes records have now authentication. Adapted to Ethereum Node record scheme.

Commit:
Nodes records have now authentication. Adapted to Ethereum Node record scheme.

Commit:
Nodes records have now authentication. Adapted to Ethereum Node record scheme.

Commit:
Nodes records have now authentication. Adapted to Ethereum Node record scheme.

mirkoAlic commented on scalanet/src/io/iohk/scalanet/crypto/package.scala:
```diff
@@ -0,0 +1,58 @@
+import org.spongycastle.crypto.params.{ECDomainParameters, ECKeyGenerationParameters, ECPrivateKeyParameters, ECPublicKeyParameters}
+import org.spongycastle.crypto.signers.{ECDSASigner, HMacDSAKCalculator}
+
+package  object crypto {
```
(minor: unnecessary space)

mirkoAlic replied:
did you copy this code from mantis right? If it that the case, i suggest you to add a comment regarding that, try to add a link to the source file where the code is.

mirkoAlic commented on scalanet/src/io/iohk/scalanet/codec/StreamObjects.scala:
```diff
@@ -0,0 +1,487 @@
+
+import scala.annotation.tailrec
+
+object IntegerStreamCodec extends StreamCodec[Int]{
```
Split this file. Check if you could re-use some of `NativeInstances` provided by decco

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/KRouterSpec.scala:
```diff
@@ -546,4 +580,6 @@ object KRouterSpec {
     import org.mockito.ArgumentMatchers.argThat
     argThat(s.contains)
   }
+
```
(minor: unnecessary spaces)

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/Generators.scala:
```diff
@@ -1,16 +1,25 @@
 object Generators {
 
-  val defaultBitLength = 16
+  val random = new SecureRandom()
```
(minor) why not called this `randomSeed`? in order to express usage intention.

mirkoAlic commented on scalanet/src/io/iohk/scalanet/crypto/package.scala:
```diff
@@ -0,0 +1,58 @@
+    val generator = new ECKeyPairGenerator
+    generator.init(new ECKeyGenerationParameters(curve, secureRandom))
+    val res = generator.generateKeyPair()
+    (res.getPrivate.asInstanceOf[ECPrivateKeyParameters],res.getPublic.asInstanceOf[ECPublicKeyParameters])
```
why you change this? 

// original one
```
  def generateKeyPair(secureRandom: SecureRandom): AsymmetricCipherKeyPair = {
    val generator = new ECKeyPairGenerator
    generator.init(new ECKeyGenerationParameters(curve, secureRandom))
    generator.generateKeyPair()
  }
```

mirkoAlic commented:
There are several places with bad formatting, try to run scalastyle or something similar in order to fix this formatting issues.

mirkoAlic commented on scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:
```diff
@@ -500,9 +519,25 @@ object KRouter {
+      s"NodeRecord(sec_number = ${sec_number}, id = ${id.toHex}, routingAddress = $routingAddress, messagingAddress = $messagingAddress)"
+  }
+
+  object NodeRecord{
```
why here?

mirkoAlic commented on scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:
```diff
@@ -500,9 +519,25 @@ object KRouter {
   // compressed public key (why)
   // TODO understand what these things do, which we need an implement.
-  case class NodeRecord[A](id: BitVector, routingAddress: A, messagingAddress: A) {
+  case class NodeRecord[A](id: BitVector, routingAddress: A, messagingAddress: A,sec_number:Long,sign:(BigInteger,BigInteger)) {
```
why here? I mean, maybe now we can move this case class, with the companion object, to a different file.

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/Generators.scala:
```diff
@@ -65,12 +74,15 @@ object Generators {
     BitVector.bits(Range(0, bitLength).map(_ => Random.nextBoolean()))
 
   def aRandomNodeRecord(
```
rename: `aRandomNodeRecord` to `genRandomNodeRecord` ?

mirkoAlic replied:
(only in case there are no so many occurrences)

mirkoAlic commented on scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:
```diff
@@ -500,9 +519,25 @@ object KRouter {
+  }
+
+  object NodeRecord{
+    def create[A](id: BitVector, routingAddress: A, messagingAddress: A,sec_number:Long,key:ECPrivateKeyParameters)(implicit codec: Codec[A]):NodeRecord[A] = {
```
I think you can redefine the `apply` method

mirkoAlic replied:
(rename: `key` to `privateKey`)

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/Generators.scala:
```diff
@@ -65,12 +74,15 @@ object Generators {
   ): NodeRecord[String] = {
-    NodeRecord(
-      id = aRandomBitVector(bitLength),
+    val pair = if(keyPair.isEmpty) crypto.generateKeyPair(random) else keyPair.get
```
val (privateKey, publicKey) = if ...

btw, instead of using Option, could we just set default value `keyPair = crypto.generateKeyPair(random)` ?

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/KPeerGroupSpec.scala:
```diff
@@ -22,7 +23,7 @@ import scala.concurrent.Await
 
 class KPeerGroupSpec extends FlatSpec {
-
+  implicit val codec:Codec[Either[NodeRecord[String],String]] = new StreamCodecFromContract[Either[NodeRecord[String], String]](new EitherCodecContract[NodeRecord[String],String](new NodeRecordCodeContract(StringCodecContract),StringCodecContract))
```
avoid having one line with this kind of size.

mirkoAlic commented on scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:
```diff
@@ -14,19 +19,23 @@ import io.iohk.scalanet.peergroup.kademlia.KRouter.{Config, NodeRecord}
 
   private val log = LoggerFactory.getLogger(getClass)
 
+  var actualNodeRecord = config.nodeRecord
```
it is safe to use var here? 

(rename: `actualNodeRecord` to `currentNodeRecord`)

mirkoAlic commented on scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:
```diff
@@ -161,17 +177,18 @@ class KRouter[A](
           pingResult <- ping(nodeToPing)
           _ <- routerState.update { current =>
-            if (pingResult) {
+            if (pingResult.isDefined && pingResult.get.sec_number >= nodeToPing.sec_number) {
```
why not `if (isFreshResult(pingResult))`

mirkoAlic replied:
Rename: `pingResult` to `nodePingResponse`

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/KRouterSpec.scala:
```diff
@@ -28,14 +28,38 @@ import scodec.bits._
   implicit override val patienceConfig: PatienceConfig =
     PatienceConfig(1 second, 100 millis)
 
+  "should accept a key with a valid signed" - {
```
"should accept a valid node"?

mirkoAlic commented on scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala:
```diff
@@ -500,9 +519,25 @@ object KRouter {
+  case class NodeRecord[A](id: BitVector, routingAddress: A, messagingAddress: A,sec_number:Long,sign:(BigInteger,BigInteger)) {
     override def toString: String =
-      s"NodeRecord(id = ${id.toHex}, routingAddress = $routingAddress, messagingAddress = $messagingAddress)"
+      s"NodeRecord(sec_number = ${sec_number}, id = ${id.toHex}, routingAddress = $routingAddress, messagingAddress = $messagingAddress)"
```
can you think for a better name than `sec_number` ? Something more meaningful..

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/KRouterSpec.scala:
```diff
@@ -28,14 +28,38 @@ import scodec.bits._
+  }
+
+  "should not accept a key with a invalid signed" - {
+    val nodeRecord =  NodeRecord[String](BitVector(Array.fill[Byte](Generators.defaultBitLength)(0)),"7643","7643",random.nextLong(),(new BigInteger(100000,random),new BigInteger(100000,random)))
```
could you create an invalid signature in a better way?

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/KRouterSpec.scala:
```diff
@@ -54,10 +78,12 @@ class KRouterSpec extends FreeSpec with Eventually {
       val krouter = aKRouter()
-      val someNodeId = aRandomBitVector()
+      val someNodeId = aRandomBitVector(264)
+      System.out.println("someNodeID: " + someNodeId)
```
remove this log line.

mirkoAlic commented on scalanet/test/src/io/iohk/scalanet/peergroup/kademlia/KRouterSpec.scala:
```diff
@@ -54,10 +78,12 @@ class KRouterSpec extends FreeSpec with Eventually {
     "should not locate an unknown node - no bootstrap" in {
       val krouter = aKRouter()
-      val someNodeId = aRandomBitVector()
+      val someNodeId = aRandomBitVector(264)
```
no magic numbers please =)

Commit:
changes in codec

Commit:
changes in codec

Commit:
Merge branch 'node_record_signed' of github.com:MaximilianoRey94/scalanet into node_record_signed

Commit:
format changes. Deleted file

Commit:
format changes. Deleted file

Commit:
format changes. Deleted file

Commit:
format changes. Deleted file

Commit:
refactoring

Commit:
refactoring

Commit:
Merge branch 'node_record_signed' of github.com:MaximilianoRey94/scalanet into node_record_signed

Commit:
delete unused class

Commit:
delete unused class

Commit:
refactoring

Commit:
refactoring

Commit:
refactoring

Commit:
refactoring

Commit:
Merge branch 'node_record_signed' of github.com:MaximilianoRey94/scalanet into node_record_signed

Commit:
add reference

Commit:
add reference

Commit:
refactoring

Commit:
refactoring

Commit:
refactoring

Commit:
refactoring

Commit:
bug fixed

Commit:
bug fixed

Commit:
refactoring

Commit:
refactoring

Commit:
modify crypto system

Commit:
modify crypto system

Commit:
modify crypto system

Commit:
modify crypto system

Commit:
modify crypto system

Commit:
modify crypto system

Commit:
# Esta es una combinación de 3 commits.
# Este es el mensaje del 1er commit:

modify crypto system

# Este es el mensaje del commit #2:

remove bad comment

# Este es el mensaje del commit #3:

refactoring

Commit:
deleted imports

Commit:
deleted imports

Commit:
deleted imports

Commit:
remove bad comment

Commit:
remove bad comment

Commit:
remove bad comment

Commit:
Merge branch 'node_record_signed' of github.com:MaximilianoRey94/scalanet into node_record_signed

Commit:
refactoring

Commit:
refactoring

Commit:
refactoring

Commit:
Merge branch 'node_record_signed' of github.com:MaximilianoRey94/scalanet into node_record_signed

Commit:
change comment

Commit:
Merge branch 'node_record_signed' of github.com:MaximilianoRey94/scalanet into node_record_signed

Commit:
refactoring
